flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FactoryUtil.java

```diff
@@ -200,6 +209,48 @@ public static TableFactoryHelper createTableFactoryHelper(
         return new TableFactoryHelper(factory, context);
     }
 
+    /**
+     * Attempts to discover an appropriate catalog factory and creates an instance of the catalog.
+     *
+     * <p>This first uses the legacy {@link TableFactory} stack to discover a matching {@link
+     * CatalogFactory}. If none is found, it falls back to the new stack using {@link Factory}
+     * instead.
+     */
+    public static Catalog createCatalog(
+            String catalogName,
+            Map<String, String> properties,
+            ReadableConfig configuration,
+            ClassLoader classLoader) {
+        // Use the legacy mechanism first for compatibility
+        try {
+            final CatalogFactory legacyFactory =
+                    TableFactoryService.find(CatalogFactory.class, properties, classLoader);
+            return legacyFactory.createCatalog(catalogName, properties);
+        } catch (NoMatchingTableFactoryException e) {
+            // No matching legacy factory found, try using the new stack
+
+            final DefaultCatalogContext context =
+                    new DefaultCatalogContext(catalogName, properties, configuration, classLoader);
```

Comment:
the `type` should not be part of the properties anymore, it should only be used for factory discovery. otherwise every factory needs to implement logic for skipping this property during validation.